src/mysql-util/src/tunnel.rs

```diff
@@ -146,49 +144,57 @@ impl Config {
-                        .into());
+                        // If the TLS configuration will validate the hostname, we need to set
+                        // the TLS hostname back to the actual upstream host and not the hostname
+                        // of the local SSH tunnel
+                        opts_builder = opts_builder.ssl_opts(Some(
+                            ssl_opts.clone().with_tls_hostname_override(Some(
+                                self.inner.ip_or_hostname().to_string(),
+                            )),
+                        ));
                     }
                 }
 
                 Ok(MySqlConn {
-                    conn: Conn::new(opts).await.map_err(MySqlError::from)?,
+                    conn: Conn::new(Into::<Opts>::into(opts_builder))
+                        .await
+                        .map_err(MySqlError::from)?,
                     _ssh_tunnel_handle: Some(tunnel),
                 })
             }
             TunnelConfig::AwsPrivatelink { connection_id } => {
                 let privatelink_host = mz_cloud_resources::vpc_endpoint_name(*connection_id);
 
-                let opts: Opts = OptsBuilder::from_opts(self.inner.clone())
-                    .ip_or_hostname(privatelink_host)
-                    .into();
+                // Override the connection host for the actual TCP connection to point to
+                // the privatelink hostname instead.
+                let mut opts_builder =
+                    OptsBuilder::from_opts(self.inner.clone()).ip_or_hostname(privatelink_host);
 
-                // TODO: Implement proper host verification when using a PrivateLink tunnel. This
-                // requires a way to modify the TLS configuration created within mysql_async
-                // to set the TLS host different from the connection host.
-                if let Some(ssl_opts) = opts.ssl_opts() {
+                if let Some(ssl_opts) = self.inner.ssl_opts() {
                     if !ssl_opts.skip_domain_validation() {
-                        return Err(anyhow!(
-                            "VERIFY_IDENTITY SSL MODE is not currently supported with PrivateLink"
-                        )
-                        .into());
+                        // If the TLS configuration will validate the hostname, we need to set
+                        // the TLS hostname back to the actual upstream host and not the
+                        // privatelink hostname.
+                        opts_builder = opts_builder.ssl_opts(Some(
+                            ssl_opts.clone().with_tls_hostname_override(Some(
+                                self.inner.ip_or_hostname().to_string(),
+                            )),
+                        ));
                     }
                 }
 
                 Ok(MySqlConn {
-                    conn: Conn::new(opts).await.map_err(MySqlError::from)?,
+                    conn: Conn::new(Into::<Opts>::into(opts_builder))
```

Comment:
What happened here? ISTM you should be able to do `Conn::new(opts_builder)`.

Reply:
Ah yes you're right! For some reason `Conn::new(opts_builder.into())` was failing to infer what type was desired which is why I switched to this, but didn't realize I could just pass the builder directly

Reply:
Ah, yeah, that would have been failing _because_ `Conn::new` accepts any type that can be turned into an `Opts`! 

```
func new_conn<O: Into<Opts>>(opts: O);

let opts_builder = OptsBuilder::new();
new_conn(opts_builder); // works because `OptsBuilder` impls `Into<Opts>`
new_conn(opts_builder.into()); // fails
```

That last line fails because Rust asks the question "what type does `opts_builder.into()` need to have?" and the answer is "O", which means "any type that can be turned into an `Opts`"—and that doesn't name a specific type! It names a whole universe of types, and Rust doesn't know which you want. When you wrote it as `Into<Opts>::into(opts_builder)`, you told Rust which specific type in that universe you wanted to use. Which was totally valid, but not as clean as just removing the superfluous `.into()`.